### PR TITLE
AD9910 driver feature extension and SUServo IIR readability

### PR DESCRIPTION
**Pull request by pmldrmota**
# ARTIQ Pull Request

## Description of Changes

The present changes extend functional features of the AD9910 driver slightly and improve the readability of the SUServo IIR filter by renaming signals and splitting statements into logically separate blocks. In anticipation of further changes, CPLD and DDS devices are passed in lists to the SUServo driver.

## Type of Changes

|   | Type |
| ------------- | ------------- |
| ✓  | :sparkles: New feature |
| ✓  | :hammer: Refactoring  |
| ✓  | :scroll: Docs |

### All Pull Requests

- [x] Use correct spelling and grammar.
- [ ] Check the copyright situation of your changes and sign off your patches (`git commit --signoff`, see [copyright](../CONTRIBUTING.rst#copyright-and-sign-off)).

### Code Changes

- [x] Run `flake8` to check code style (follow PEP-8 style). `flake8` has issues with parsing Migen/gateware code, ignore as necessary.
- [x] Test your changes or have someone test them. Mention what was tested and how.
- [x] Add and check docstrings and comments
- [x] Check, test, and update the [unittests in /artiq/test/](../artiq/test/) or [gateware simulations in /artiq/gateware/test](../artiq/gateware/test)

### Documentation Changes

- [x] Check, test, and update the documentation in [doc/](../doc/). Build documentation (`cd doc/manual/; make html`) to ensure no errors.

### Git Logistics

- [x] Split your contribution into logically separate changes (`git rebase --interactive`). Merge/squash/fixup commits that just fix or amend previous commits. Remove unintended changes & cleanup. See [tutorial](https://www.atlassian.com/git/tutorials/rewriting-history/git-rebase).
- [x] Write short & meaningful commit messages. Review each commit for messages (`git show`). Format:
  ```
  topic: description. < 50 characters total.
  
  Longer description. < 70 characters per line
  ```

### Licensing

See [copyright & licensing for more info](https://github.com/m-labs/artiq/blob/master/CONTRIBUTING.rst#copyright-and-sign-off).
ARTIQ files that do not contain a license header are copyrighted by M-Labs Limited and are licensed under LGPLv3+.

### CODE REVIEW

File: artiq/coredevice/ad9910.py [hunk 1]
```diff
@@ -242,6 +242,10 @@ def read32(self, addr):
 
         :param addr: Register address
         """
+        return self.read32_impl(addr)
+
+    @kernel
+    def read32_impl(self, addr):
```
Comment (dnadlinger) [inline artiq/coredevice/ad9910.py:248]: This should probably in the (to come) suservo commit, with an explanatory comment.

**Comment by pmldrmota** [comment]
Can this be pulled or are there objections? I have follow-up PRs prepared based on those changes to approach phase-coherent SUServo step by step.

**Comment by jordens** [comment]
There needs to be a review before it can be merged. I haven't had time to do it yet. Maybe this week.
Please do describe what level of testing (hardware, unittests, docs) you have done.

**Comment by pmldrmota** [comment]
Sure! 

- @dnadlinger and I looked at the effect of the phase_clear bit in CFR1 by first setting FTW=0, then setting phase_clear and observing the DDS output after setting FTW>0. We ended up using the phase_autoclear now, but the option to set this bit using the function set_cfr1 remains valid.

- Introducing the set_cfr2 function is a non-functional change and is tested over 4 months now in one of our setups. It just improves readability.

- After the non-functional suservo gateware refactoring, all artiq/gateware/test scripts pass without error.

- The docs build without problems on my PC.

**Comment by jordens** [comment]
Never mind the merits or reasons for the changes: apart from building the docs and the gateware tests, did you test **this** (not another branch that includes these and other changes) code on hardware?

**Comment by pmldrmota** [comment]
No, this branch was not tested in isolation.

**Comment by jordens** [comment]
What's the plan? I don't think we can offload the testing of your PR to users of the ARTIQ rolling beta.
Alternatively, we can review it untested and then merge into a staging branch and accumulate your other PRs there until you feel ready for testing. It may just make debugging harder.

**Comment by pmldrmota** [comment]
I think that's a good idea!

File: artiq/coredevice/suservo.py [hunk 2]
```diff
@@ -57,32 +57,26 @@ class SUServo:
...
     """
-    kernel_invariants = {"channel", "core", "pgia", "cpld0", "cpld1",
-                         "dds0", "dds1", "ref_period_mu"}
+    kernel_invariants = {"channel", "core", "pgia", "cplds", "ddses",
+                         "ref_period_mu"}
 
     def __init__(self, dmgr, channel, pgia_device,
-                 cpld0_device, cpld1_device,
-                 dds0_device, dds1_device,
+                 cpld_devices, dds_devices,
```
Comment (dnadlinger) [inline artiq/coredevice/suservo.py:71]: Need a changelog (`RELEASE_NOTES.rst`) entry for these API changes.

Also, are calls generated from any of the JSON device database tooling?

**Comment by dnadlinger** [comment]
I've mad some small changes and retargeted this back to m-labs/master, as it is high time things started moving here. Keep in mind that we have been using our changes for the better part of a year on all setups now, including applications where we'd really notice any glitches (phase-coherent suservo for optical qubit gate beams).

Are there no HITL tests for suservo?

**Comment by sbourdeauducq** [comment]
> As for further testing of this in isolation, are there no HITL tests for suservo?

Currently we don't have them, could be added though.

**Comment by sbourdeauducq** [comment]
Should we merge this?

**Comment by cjbe** [comment]
>Should we merge this?

Is there any reason not to merge this now @pmldrmota ?

**Comment by pmldrmota** [comment]
Of course this can be merged.
This PR is only a preparatory cleanup with no real benefit if the follow-up PRs do not follow in tight succession. 
I am willing to prepare the next PR for the full multi-channel phase-coherent SUServo upgrade (essentially updating https://github.com/pmldrmota/artiq/tree/suservo-multi-phasecoherent) if this is the general interest.

**Comment by cjbe** [comment]
>I am willing to prepare the next PR for the full multi-channel phase-coherent SUServo upgrade ... if this is the general interest.

I think this would be of general interest - I have had quite a few groups ask how we are managing using SUServo on our coherent beams.

**Pull request merged**